packages/react-reconciler/src/__tests__/ReactOffscreen-test.js

```diff
@@ -1771,5 +1833,252 @@ describe('ReactOffscreen', () => {
+
+    await act(async () => {
+      root.render(<App />);
+    });
+
+    expect(Scheduler).toHaveYielded([
+      'outer',
+      'middle',
+      'inner',
+      'mount layout inner',
+      'mount layout middle',
+      'mount layout outer',
+      'mount inner',
+      'mount middle',
+      'mount outer',
+    ]);
+
+    expect(outerOffscreen).not.toBeNull();
+    expect(innerOffscreen).not.toBeNull();
+
+    await act(async () => {
+      outerOffscreen.detach();
+    });
+
+    expect(innerOffscreen).toBeNull();
+
+    expect(Scheduler).toHaveYielded([
+      'unmount layout middle',
+      'unmount layout inner',
+      'unmount middle',
+      'unmount inner',
+    ]);
+
+    await act(async () => {
+      outerOffscreen.attach();
+    });
+
+    expect(Scheduler).toHaveYielded([
+      'mount layout inner',
+      'mount layout middle',
+      'mount inner',
+      'mount middle',
+    ]);
+
+    await act(async () => {
+      innerOffscreen.detach();
+    });
+
+    expect(Scheduler).toHaveYielded(['unmount layout inner', 'unmount inner']);
+
+    // Calling detach on already detached Offscreen.
+    await act(async () => {
+      innerOffscreen.detach();
+    });
+
+    expect(Scheduler).toHaveYielded([]);
+
+    await act(async () => {
+      innerOffscreen.attach();
+    });
+
+    expect(Scheduler).toHaveYielded(['mount layout inner', 'mount inner']);
+
+    await act(async () => {
+      innerOffscreen.detach();
+      outerOffscreen.attach();
+    });
+
+    expect(Scheduler).toHaveYielded(['unmount layout inner', 'unmount inner']);
+  });
+
+  // @gate enableOffscreen
+  it('batches multiple attach and detach calls scheduled from an event handler', async () => {
+    function Child() {
+      useEffect(() => {
+        Scheduler.unstable_yieldValue('attach child');
+        return () => {
+          Scheduler.unstable_yieldValue('detach child');
+        };
+      }, []);
+      return 'child';
+    }
+
+    const offscreen = React.createRef(null);
+    function App() {
+      return (
+        <Offscreen ref={offscreen} mode="manual">
+          <Child />
+        </Offscreen>
+      );
+    }
+
+    const root = ReactNoop.createRoot();
+    await act(() => {
+      root.render(<App />);
+    });
+
+    expect(Scheduler).toHaveYielded(['attach child']);
+
+    await act(async () => {
+      const instance = offscreen.current;
+      // Detach then immediately attach the instance.
+      instance.detach();
+      instance.attach();
+    });
+
+    expect(Scheduler).toHaveYielded([]);
+
+    await act(async () => {
+      const instance = offscreen.current;
+      instance.detach();
+    });
+
+    expect(Scheduler).toHaveYielded(['detach child']);
+
+    await act(async () => {
+      const instance = offscreen.current;
+      // Attach then immediately detach.
+      instance.attach();
+      instance.detach();
+    });
+
+    expect(Scheduler).toHaveYielded(['attach child', 'detach child']);
```

Comment:
attach is synchronous, that's why child is attached and then immediately detached.

Reply:
It should still be batched, though. That's how a discrete update works, for example. Or a setState inside a layout effect.

The model should be that it has the same behavior as a setState. Any clever tricks we do on top of that are only optimizations — like skipping the render phase if we can prove it's not necessary.